tests/test_cases/issue_751/issue_751.py

```diff
@@ -0,0 +1,72 @@
+"""Test for comparing handle classes
+
+https://github.com/cocotb/cocotb/pull/751
+"""
+
+import cocotb
+from cocotb.handle import SimHandleBase, NonHierarchyObject
+from cocotb.triggers import RisingEdge, FallingEdge
+from cocotb.clock import Clock
+
+class Testbench:
+    def __init__(self, dut):
+        self.dut = dut
+        self.clkedge = RisingEdge(dut.clk)
+
+    @cocotb.coroutine
+    def initialise(self):
+        cocotb.fork(Clock(self.dut.clk, 10).start())
+        self.dut.reset <= 0
+        for _ in range(2):
+            yield self.clkedge
+        self.dut.reset <= 1
+        for _ in range(3):
+            yield self.clkedge
+
+
+@cocotb.test()
+def test_compare_simhandlebase(dut):
+    """Test for SimHandleBase comparisons"""
+    tb = Testbench(dut)
+    yield tb.initialise()
+    for _ in range(3):
+        yield tb.clkedge
+
+    # Want to check the __eq__ comparator in SimHandleBase
+    # (overridden in NonHierarchyObject)
+    assert isinstance(dut.i_module_a, SimHandleBase)
+    assert not isinstance(dut.i_module_a, NonHierarchyObject)
+    assert isinstance(dut.i_module_b, SimHandleBase)
+    assert not isinstance(dut.i_module_b, NonHierarchyObject)
+
+    # Same handle
+    assert dut.i_module_a.__eq__(dut.i_module_a)
+    # Different handles
+    assert not dut.i_module_a.__eq__(dut.i_module_b)
+    # Compare against non-handle not implemented
+    assert dut.i_module_a.__eq__(1) == NotImplemented
+
+
+@cocotb.test()
+def test_compare_nonhierarchy(dut):
+    """Test for NonHierarchyObject comparisons"""
+    tb = Testbench(dut)
+    yield tb.initialise()
+    for _ in range(3):
+        yield tb.clkedge
+
+    # Check that all these signals are NonHierarchyObject children
+    assert isinstance(dut.counter_plus_two, NonHierarchyObject)
+    assert isinstance(dut.counter_plus_five, NonHierarchyObject)
+    assert isinstance(dut.clk, NonHierarchyObject)
+    assert isinstance(dut.i_module_a.clk, NonHierarchyObject)
+
+    # Two different handles
+    assert not dut.counter_plus_two.__eq__(dut.counter_plus_five)
+    # Two different handles with the same value
+    # Because they are handles, it is checked if they are the same handle
+    assert not dut.clk.__eq__(dut.i_module_a.clk)
+    # A handle and a value
+    # Because one is a value, it is compared against the value of the handle
+    yield tb.clkedge
+    assert dut.clk.__eq__(1)
```

Comment:
I'd be inclined to assert these with `==`

Reply:
What about the one where I'm checking for `NotImplemented`?

Reply:
Well this particular line is wrong, because both `assert True` and `assert NotImplemented` will pass.

The NotImplemented checks could be handled by comparing against objects like these:
```python
class _EqualsEverything:
    def __eq__(self, other):
        return True
    def __ne__(self, other):
        return False
EqualsEverything = _EqualsEverything()
class _EqualsNothing:
    def __eq__(self, other):
        return False
    def __ne__(self, other):
        return True
EqualsNothing = _EqualsNothing()
EqualsNotImplemented = object()  # this one is builtin
```
Used as
```python
def test_equality_delegates_if_unknown():
    # note: the dummy class goes on the right to ensure that the Handle method is dispatched first
    assert dut.clk == EqualsEverything
    assert not (dut.clk == EqualsNothing)
    assert not (dut.clk == EqualsNotImplemented)

    # always test the inverse operators too
    assert not (dut.clk != EqualsEverything
    assert (dut.clk != EqualsNothing)
    assert (dut.clk != EqualsNotImplemented)
```

Reply:
That's good to know, will change these asserts then